[RFC] Use GNULIB's compiler warning code

fidencio opened the pull request:
This patch series was sent to the sssd-devel and some discussions
already happened there[0](https://lists.fedorahosted.org/archives/list/sssd-devel@lists.fedorahosted.org/thread/CTGR5CYV2PT3PLFLIRBCRTJP5ZW2XBMO/#CTGR5CYV2PT3PLFLIRBCRTJP5ZW2XBMO). I've decided to open the PR because there
are some few patches that can be pushed even if we decide to not use
the "many warnings" patches.

Let's keep track of those patches (and discussions related to them) in
the github, in this way we can avoid them to get lost. :-)

Best Regards,

Changes:

683f72d (Fabiano Fidêncio, 10 weeks ago)
   BUILD: Make use of GNULIB's compiler warning code

   As GNULIB has the 'manywarnings' module, which basically turns on every GCC
   warning, let's make use of it. We can easily blacklist the warnings we
   cannot cope with, but the main goal should be to have enabled every
   possible GCC warning.

   When new GCC warnings are created the 'manywarnings' file can be refreshed
   from upstream GNULIB.

   Signed-off-by: Fabiano Fidêncio fidencio@redhat.com

f59828a (Fabiano Fidêncio, 5 days ago)
   NSS: Fix "old-style-definition" warning caught by GCC

   Signed-off-by: Fabiano Fidêncio fidencio@redhat.com

f22aff7 (Fabiano Fidêncio, 5 days ago)
   SIFP: Fix a "jump-misses-init" warning caught by GCC

   Signed-off-by: Fabiano Fidêncio fidencio@redhat.com

58609d3 (Fabiano Fidêncio, 5 days ago)
   RESOLV: Fix a "-Werror=null-dereference" caught by GCC

   Signed-off-by: Fabiano Fidêncio fidencio@redhat.com

bd1d7fd (Fabiano Fidêncio, 5 days ago)
   RESOLV: Simplify reply_weight_rearrange() a little bit

   Signed-off-by: Fabiano Fidêncio fidencio@redhat.com

lslebodn commented:
It seems that resolv patches broke  `resolv-tests` on fedora 22 and fedora 23.
The test has been running for 30 hours without any result. I've just cancelled it.

BTW CI did not passed on any supported distro.

fidencio commented:
I've updated this series and I'm still waiting for CI to finish.
A lot of warnings were just caught when building on RHEL6 machine (which passes gracefully now).

I'll post the CI results as soon as I have them.

fidencio commented:
http://sssd-ci.duckdns.org/logs/job/56/78/summary.html

lslebodn commented:
I do not think that initialisation to NULL is the best solution for Wunitialized on RHLE6.

My experience is that such bugs occurs when there is some suspicious code in functions which are
inlined.

lslebodn commented on src/providers/krb5/krb5_child.c:
```diff
@@ -1464,7 +1464,7 @@ static errno_t changepw_child(struct krb5_req *kr, bool prelim)
                                         prompter, kr, 0,
-                                        SSSD_KRB5_CHANGEPW_PRINCIPAL,
+                                        discard_const(SSSD_KRB5_CHANGEPW_PRINCIPAL),
                                         kr->options);
```
Such patch was rejected by simo approximately 3 years ago.
Because it was fixed in krb5.

fidencio replied:
Okay, NACK to this one.

lslebodn commented on src/tests/sbus_codegen_tests.c:
```diff
@@ -115,15 +115,15 @@ END_TEST
 START_TEST(test_signals)
 {
-    const struct sbus_signal_meta *signal;
+    const struct sbus_signal_meta *sm_signal;
```
What does the prefix "sm_" mean?

fidencio replied:
"sm_" comes from sbus_meta.

lslebodn replied:
It is not `sbus_meta_signal` but `sbus_signal_meta` But `ssm_` does not increase readability either

What about `a_signal` or `sig`, `signl`

fidencio replied:
Okay, I'll follow one of your suggestions.

lslebodn commented on src/tools/sssctl/sssctl_cache.c:
```diff
@@ -31,10 +31,10 @@
+#define SSSCTL_CACHE_UPDATE {_("Cache entry last update time"), SYSDB_LAST_UPDATE, attr_time_fn}
+#define SSSCTL_CACHE_EXPIRE {_("Cache entry expiration time"), SYSDB_CACHE_EXPIRE, attr_expire_fn}
+#define SSSCTL_CACHE_IFP    {_("Cached in InfoPipe"), SYSDB_IFP_CACHED, attr_yesno_fn}
 #define SSSCTL_CACHE_NULL   {NULL, NULL, NULL}
```
Suffix "_fn" in name of function looks weird to me.
I would suggest either rename function (get_attr_name ...) to have more suitable name
or rename local variables which shadow functions.

fidencio replied:
Okay, added the get_ prefix to these functions.

lslebodn commented on src/providers/ad/ad_gpo.c:
```diff
@@ -100,7 +100,6 @@
 /* If INI_PARSE_IGNORE_NON_KVP is not defined, use 0 (no effect) */
 #ifndef INI_PARSE_IGNORE_NON_KVP
 #define INI_PARSE_IGNORE_NON_KVP 0
-#warning INI_PARSE_IGNORE_NON_KVP not defined.
```
NACK,
this warning is intentional. It was added because sssd need to be compiled even with libini_config-1.2.
Because required the latest libini_config-1.3 is only in fedora and is required for
https://fedorahosted.org/sssd/ticket/2751

fidencio replied:
Fair enough ...

fidencio commented:
@lslebodn, a way easier solution would be to disable -Werror when running CI on RHEL6 than fixing those issues only caught there. Is there a simple way to do that?

lslebodn commented on src/tools/tools_util.c:
```diff
@@ -73,18 +73,17 @@ static int setup_db(struct tools_ctx *ctx)
         }
 
-        fprintf(stderr, "%s%s", error, nl);
+        fprintf(stderr, "%s", error);
```
ACK

lslebodn commented on src/tests/krb5_child-test.c:
```diff
@@ -210,8 +210,18 @@ create_dummy_req(TALLOC_CTX *mem_ctx, const char *user,
+        DEBUG(SSSDBG_FATAL_FAILURE,
+              "Failed to create dummy pam_data");
+        goto fail;
+    }
```
LGTM,

lslebodn commented:
On (09/11/16 07:54), fidencio wrote:
>@lslebodn, a way easier solution would be to disable -Werror when running CI on RHEL6 than fixing those issues only caught there. Is there a simple way to do that?
>
I agree that it would be better to disable Werror on rhel6.

But your patch "BUILD: Make use of GNULIB's compiler warning code "
introduced Werror. And I am still not satisfied with the patch
because of increased time in CI (10%) caused by detection in manywarnings.

LS

fidencio commented:
Disabling Werror on RHEL6 will be just a matter of passing "--enable-werror=no" to configure. It should be **really** simple as long as we agree on doing that.

About the extra extra time, it was 5% last time you reviewed. 100 seconds for a process that we already have to wait for an hour or so. IMO should not block, at all, introducing more warnings to our code.

lslebodn commented:
On (09/11/16 08:18), fidencio wrote:
>Disabling Werror on RHEL6 will be just a matter of passing "--enable-werror=no" to configure. It should be **really** simple as long as we agree on doing that.
>
>About the extra extra time, it was 5% last time you reviewed. 100 seconds for a process that we already have to wait for an hour or so. IMO should not block, at all, introducing more warnings to our code.
>
I am not against enabling new warnings.
More warnings can be enabled in Makefile.am directly.
The warnings which are not supported by older compilers (gcc on rhel6)
should be explicitely checked at configure time.

It is wasting of time to check evey time flags which
does not produce any warnings atm:
e.g.
-Wabi
-Warray-bounds
-Wcast-align
-Wcast-qual
-Wclobbered
-Wdisabled-optimization
-Wempty-body
-Werror-implicit-function-declaration
-Wformat-security
-Wignored-qualifiers
-Winit-self
-Winvalid-pch
-Wlarger-than=65536
-Wmissing-include-dirs
-Wmissing-parameter-type
-Wmultichar
-Wnested-externs
-Wold-style-declaration
-Woverlength-strings
-Woverride-init
-Wpacked
-Wpointer-arith
-Wshadow
-Wstack-protector
-Wstrict-prototypes
-Wsync-nand
-Wtype-limits
-Wundef
-funsafe-loop-optimizations
-Wunsafe-loop-optimizations
-Wunused-but-set-parameter
-Wwrite-strings
-Wall

Or warnings which will never be fixed:
e.g.
-Wsystem-headers
-Wformat-nonliteral
-Wpacked
-Wvla
-Wtraditional

LS

fidencio commented:
Lukáš,

I'm afraid you didn't understand how the usage of the GNULIB's compiler warnings is supposed to be.
There are basically 3 essential files:
- warnings.m4
- manywarnings.m4
- sssd-compile-warnings.m4

The first two you can just get from GNULIB's git repo and update in our repo whenever there's a update there. So, tweaking those files is something I could do, but it's wrong per si.
There's the third file then, sssd-compile-warnings.m4, where we disable the warnings we don't need, want or can't cope with. And in this file I've added all the warnings we don't need, want or can't cope with. In this very same file, we check whether a warning is or is not supported.

That's how it's supposed to work. If you think patching Makefile.am is better, okay, just reject this patch series, open a bug and we will mark as "patches welcome".

Again, I really don't see a valid point for rejecting this series because it introduces 100s more in a process that takes an hour or so.

So, what's your suggestion? Can we accept the series? Can you, at least, finish reviewing the patches that are not RHEL6 specific?

fidencio commented:
@lslebodn:
After some proper checking, seems that all "-Wunitialized" reports are false positives, so I've removed the patches from this series.
I've added "Reviewed-by: " to the patches you've already ACKed and dropped the ones you've already NACKed. For the ones you gave some suggestions, I followed them.

The build is failing in CentOS CI and I'll take a look on it later on. I've updated my branch based only on my local tests (and they're okay on F25). I haven't done any test on our CI because it seems to be broken due to lack of space.

Feel free to wait till I solve those issues before starting reviewing. Or feel free to NACK the series as soon as possible, so I don't spend time on it anymore.

fidencio commented:
So, here are some stats about the configure time that you complained before:
So, I've ran the contrib/ci/run script on our F25 machine on Jenkins.

This is what we have **without**  GNULIB patch:
```
-bash-4.3$ ./contrib/ci/run
install-deps:           success  00:00:05 ci-install-deps.log
autoreconf:             success  00:00:22 ci-autoreconf.log
DEBUG BUILD:                              ci-build-debug
configure:              success  00:00:16 ci-build-debug/ci-configure.log
make-tests:             success  00:02:55 ci-build-debug/ci-make-tests.log
make-check-valgrind:    success  00:01:33 ci-build-debug/ci-make-check-valgrind.log
SUCCESS

```

And this is what we have **with** GNULIB patch applied:
```
-bash-4.3$ ./contrib/ci/run 
install-deps:           success  00:00:05 ci-install-deps.log
autoreconf:             success  00:00:23 ci-autoreconf.log
DEBUG BUILD:                              ci-build-debug
configure:              success  00:00:22 ci-build-debug/ci-configure.log
make-tests:             success  00:03:56 ci-build-debug/ci-make-tests.log
make-check-valgrind:    success  00:01:33 ci-build-debug/ci-make-check-valgrind.log
SUCCESS

```

It adds 6 seconds more in the configure time. I really think that dropping those patches because of 6 seconds on the configure is far, but really really, far away from being reasonable.

So, can we have the patches reviewed and merged?

fidencio commented:
It's been almost 2 months this code has been around and there's anyone interested to review.
So, I'm closing the PR (as "rejected") as this patchset doesn't seem to be useful for the project.

fidencio commented:
So, I've re-worked those patches a little bit and here is the time difference when running reconfing with the patches:
```
real	0m26.047s
user	0m21.318s
sys	0m4.635s
```

And now without:
```
real	0m25.565s
user	0m20.696s
sys	0m4.433s
```

This patch set is rebased on top of PR #377.

I really would appreciate if someone could review and give their opinion.

The reason this PR was blocked is because this time difference has been considered a "performance issue".

@jhrozek, could you take a look on this?

fidencio commented:
Ah, somehow seems that I cannot reopen the PR. So, I'm opening a new one.
